pkg/controller/machine.go

```diff
@@ -562,45 +562,60 @@ func (c *controller) machineDelete(machine *v1alpha1.Machine, driver driver.Driv
 			}
 
 			err = driver.Delete()
+			if err != nil {
+				// When machine deletion fails
+				glog.Errorf("Deletion failed for machine %q: %s", machine.Name, err)
+
+				lastOperation := v1alpha1.LastOperation{
+					Description:    "Cloud provider message - " + err.Error(),
+					State:          v1alpha1.MachineStateFailed,
+					Type:           v1alpha1.MachineOperationDelete,
+					LastUpdateTime: metav1.Now(),
+				}
+				currentStatus := v1alpha1.CurrentStatus{
+					Phase:          v1alpha1.MachineFailed,
+					TimeoutActive:  false,
+					LastUpdateTime: metav1.Now(),
+				}
+				c.updateMachineStatus(machine, lastOperation, currentStatus)
+				return err
+			}
 
-			// Delete node object if exists
+			// Delete node object if exists delete it
 			if nodeName != "" {
 				err = c.targetCoreClient.CoreV1().Nodes().Delete(nodeName, &metav1.DeleteOptions{})
 				if err != nil && !apierrors.IsNotFound(err) {
+					message := fmt.Sprintf("Deletion of Node Object %q failed due to error: %s", nodeName, err)
+					lastOperation := v1alpha1.LastOperation{
+						Description:    message,
+						State:          v1alpha1.MachineStateFailed,
+						Type:           v1alpha1.MachineOperationDelete,
+						LastUpdateTime: metav1.Now(),
+					}
+					c.updateMachineStatus(machine, lastOperation, machine.Status.CurrentStatus)
```

Comment:
machine might have been deleted succesfully [here](https://github.com/gardener/machine-controller-manager/pull/324/files#diff-d8287fe74b5273163c9f6b6c635ad912R564). So this step would fail.

It is much easier to just add check `if nodeName != "" ` before deleting the node in the code before #322 was merged.

Reply:
I would suggest revert #322 and just add a the check mentioned above.

Reply:
Only the VM is deleted above. The machine object still remains below right.

Reply:
Yes. Realised this just after posting my comments. :)